graphql_compiler/schema_transformation/split_query.py

```diff
@@ -34,6 +46,478 @@ def __init__(self, query_ast):
+                intra_schema_fields.append(vertex_field)
+        else:
+            raise AssertionError(
+                u'Input vertex field {} is not a Field'.format(vertex_field)
+            )
+    return intra_schema_fields, cross_schema_fields
+
+
+def _get_selections_from_property_and_vertex_fields(property_fields_map, vertex_fields):
+    """Combine property fields and vertex fields into a list of selections.
+
+    Args:
+        property_fields_map: OrderedDict[str, Field], mapping name of field to their
+                             representation. It is not modified by this function
+        vertex_fields: List[Field]. It is not modified by this function
+
+    Returns:
+        List[Field], containing all property fields then all vertex fields, in order
+    """
+    selections = list(six.itervalues(property_fields_map))
+    selections.extend(vertex_fields)
+    return selections
+
+
+def _get_child_query_node_and_out_name(ast, child_type_name, child_field_name, name_assigner):
+    """Create a query node out of ast, return node and unique out_name on field with input name.
+
+    Create a new document out of the input AST, that has the same structure as the input. For
+    instance, if the input AST can be represented by
+        out_Human {
+          name
+        }
+    where out_Human is a vertex field going to type Human, the resulting document will be
+        {
+          Human {
+            name
+          }
+        }
+    If the input AST starts with a type coercion, the resulting document will start with the
+    coerced type, rather than the original union or interface type.
+
+    The output child_node will be wrapped around this new Document. In addition, if no field
+    of child_field_name currently exists, such a field will be added. If there is no @output
+    directive on this field, a new @output directive will be added.
+
+    Args:
+        ast: Field or InlineFragment, representing the AST that we're using to build a child
+             node. It is not modified by this function
+        child_field_name: str. If no field of this name currently exists as a part of the root
+                          selections of the input AST, a new field will be created in the AST
+                          contained in the output child query node
+        name_assigner: IntermediateOutNameAssigner, object used to generate and keep track of
+                       names of newly created @output directives
+
+    Returns:
+        Tuple[SubQueryNode, str], the child sub query node wrapping around the input AST, and
+        the out_name of the @output directive uniquely identifying the field used for stitching
+        in this sub query node
+    """
+    # Get type and selections of child AST, taking into account type coercions
+    child_selection_set = ast.selection_set
+    if (
+        len(child_selection_set.selections) == 1 and
+        isinstance(child_selection_set.selections[0], InlineFragment)
+    ):
```

Comment:
Consider making a helper function for this check, something like `_has_single_fragment_selection` maybe -- possibly even better if it returns the fragment selection AST so you can avoid the `[0]` below. It seems to be repeated a few times.

Maybe you could even use the `_get_inline_fragment(ast)` function from `compiler_frontend.py`?

Reply:
`_get_inline_fragment` does the right thing, although it raises an GraphQLCompilationError which I don't like very much -- would prefer GraphQLValidationError. You're definitely right I'm repeating myself here somewhat, I'll use that function for now.